File: Top-Interview-Questions/LinkedList/19-remove-nth-node-fromend-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        first = second = head
        
        for f in range(n):
            first=first.next
           # print("firstval", first.val)
        if first is None:
            return head.next
        while first.next:
            first=first.next
           # print("firstval2", first)
            second=second.next
           # print("secondval", second.val)
           # print("secondval2", second.next.val)
           # print("secondval3", second.next.next)
       # print("final first val", first)
       # print("final second val", second)
        second.next = second.next.next
        return head
    # TC: O(n), where n is the length of the linked list
    # SC: O(1), no additional space
    # Runtime: 32 ms, faster than 68.75%
    # Memory: 13.7 MB, less than 6.06%

File: Top-Interview-Questions/LinkedList/19-remove-nth-node-fromend-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Time Complexity: O(head-n)
// Space Complexity: O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 10.5 MB, less than 5.26%
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        if(head->next == NULL)
            return NULL;
        
        ListNode* tail = head;
        
        //remove tail
        if(n == 1){
            while(tail->next->next != NULL){
                tail = tail->next;
            }
            tail->next = NULL;
            return head;
        }
        
        int count = 0;
        
        //count distance to tail from head
        while(tail->next != NULL){
            count++;
            tail = tail->next;
        }
        
        //if head, return head->next pointer
        if(count == n-1){
            head = head->next;
            return head;
        }
        
        tail = head;
        
        //go to right before to the node that you wish to delete
        for(int i = 0; i < count-n; i++){
            tail = tail->next;
        }        
        
        tail->next = tail->next->next;
        
        return head;
    }
};